chess-AI/train.py

```diff
@@ -15,73 +23,144 @@ class MctsTrain:
-                                                                                search_probs,
-                                                                                board)
-            
-            # Adds entry to the training examples
-            self.training_examples.append([fen_string, full_search_probs, None])
+            for _ in range(epochs):
+                # The board needs to be initialized to the starting state
+                board = chess.Board()
 
-            # Makes the random action on the board, and gets fen string
-            move = chess.Move.from_uci(move)
-            board.push(move)
-            fen_string = board.fen()
+                while True:
+                    fen_string = board.fen()
 
-            # if the game is over then that is the end of the episode
-            if board.is_game_over():
-                # Need to assign the rewards to the examples
-                self.assign_rewards(board)
-                return
+                    # Perform mcts simulations
+                    for _ in range(self.mcts_simulations):
+                        self.mcts.search(board, nnet)
 
+                    # Gets the moves and policy from the mcts, as well as the individual move to take
+                    # NOTE: moves[i] corresponds to search_probs[i]
+                    moves, search_probs, move = self.mcts.find_search_probs(fen_string, 5)
+            
+                    # Store board state (used in training loop)
+                    self.boards.append(fen_string)
+
+                    # Converts mcts search probabilites to (8,8,73) vector
+                    full_search_probs = self.policy_converter.compute_full_search_probs(moves, search_probs, board)
+                    
+                    # Adds entry to the training examples
+                    self.mcts_probs.append(full_search_probs)
+                    self.mcts_evals.append(None)
+
+                    # Makes the random action on the board, and gets fen string
+                    move = chess.Move.from_uci(move)
+                    board.push(move)
+                    print(board)
+
+                    # If the game is over, end the episode
+                    if board.is_game_over() or board.is_stalemate() or board.is_seventyfive_moves() or board.is_fivefold_repetition() or board.can_claim_draw():
+                        self.assign_rewards(board)
+                        break
+
+                self.mcts_probs = torch.tensor(np.array(self.mcts_probs)).float()
+                self.mcts_evals = torch.tensor(self.mcts_evals).float()
+
+                # Create iterable dataset with mcts labels
+                dataset = TensorDataset(self.mcts_probs, self.mcts_evals)
+                train_dl = DataLoader(dataset=dataset, batch_size=batch_size, shuffle=False)
+
+                # Define loss functions
+                loss_fn1 = torch.nn.CrossEntropyLoss()
```

Comment:
Pls change to `ce_loss_fn` and `mse_loss_fn` for readability

Reply:
You can then optionally make a list `loss_fns = [ce_loss_fn, mse_loss_fn]` and iterate over this list to compute loss for all the different loss functions you care about